### PR TITLE
fixes #1086 - Disable Zookeeper retries from TabletServer

### LINKED ISSUE #1086: ZooLock uses ZooReader which may retry on session expired
Accumulo tablet servers use ephemeral nodes in zookeeper to advertise their existence. If this ephemeral node is deleted for any reason a tablet server must halt.   One way it can be deleted is if  a tablet server is not responsive enough to zookeeper servers.  In this case the zookeeper servers will close the session which causes any ephemeral nodes related to the session to be deleted.  If the tablet server is still alive and tries to communicate with zookeeper it will get a session expired exception.  When this happens the tablet server should halt.

The code that manages tablet servers ephemeral nodes is called ZookLock.  ZooLock uses an Accumulo abstraction  for Zookeeper called ZooReader.  The ZooReader code may retry on session expired events.  Need to examine ZooLock's use of ZooReader to ensure that retrying on session expired events does not delay ZooLock from becoming aware of session expiration.

### PR DESCRIPTION
**Pull request by dlmarion**
Looking at the issues at https://github.com/apache/accumulo/issues/1086#issuecomment-670084908, I believe that just disabling the retries from the server components resolves both. This change disables retries from the server components and uses the default retry factory when being called from non-Server classes.

### CODE REVIEW

File: core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooLock.java [hunk 1]
```diff
@@ -69,13 +70,14 @@
   private boolean watchingParent = false;
   private String asyncLock;
 
-  public ZooLock(ZooReaderWriter zoo, String path) {
+  public ZooLock(ZooReaderWriter zoo, String path, RetryFactory retryFactory) {
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooLock.java:73]: A new parameter is added, but not used.

File: core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooReader.java [hunk 2]
```diff
@@ -38,24 +38,36 @@
...
   private static final Logger log = LoggerFactory.getLogger(ZooReader.class);
 
-  protected static final RetryFactory RETRY_FACTORY = Retry.builder().maxRetries(10)
+  public static final RetryFactory DEFAULT_RETRY_FACTORY = Retry.builder().maxRetries(10)
+      .retryAfter(250, MILLISECONDS).incrementBy(250, MILLISECONDS).maxWait(5, TimeUnit.SECONDS)
+      .backOffFactor(1.5).logInterval(3, TimeUnit.MINUTES).createFactory();
+
+  public static final RetryFactory DISABLED_RETRY_FACTORY = Retry.builder().maxRetries(0)
       .retryAfter(250, MILLISECONDS).incrementBy(250, MILLISECONDS).maxWait(5, TimeUnit.SECONDS)
       .backOffFactor(1.5).logInterval(3, TimeUnit.MINUTES).createFactory();
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooReader.java:47]: Can probably remove all this extra retry configuration since the max retries is zero here. I don't think these are all required parameters are they?

File: core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooReaderWriter.java [hunk 3]
```diff
@@ -49,13 +49,14 @@
   public ZooReaderWriter(AccumuloConfiguration conf) {
     this(conf.get(Property.INSTANCE_ZK_HOST),
         (int) conf.getTimeInMillis(Property.INSTANCE_ZK_TIMEOUT),
-        conf.get(Property.INSTANCE_SECRET));
+        conf.get(Property.INSTANCE_SECRET), ZooReader.DISABLED_RETRY_FACTORY);
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooReaderWriter.java:52]: The behavioral differences between the constructors is unintuitive. Can we instead use a well-named static factory method, instead of the constructor to construct a ZRW for the specific intended purpose, so the constructor isn't unintentionally used by other code, and subject to the non-retrying behavior accidentally?

```java
// e.g.
ZooReaderWriter zrw1 = ZooReaderWriter.retrying(conf);
ZooReaderWriter zrw2 = ZooReaderWriter.nonRetrying(conf);
// or
ZooReaderWriter zrw = ZooReaderWriter.forXuseOnly(conf);
```

File: server/manager/src/main/java/org/apache/accumulo/master/Master.java [hunk 4]
```diff
@@ -1422,7 +1423,8 @@ private void getMasterLock(final String zMasterLoc) throws KeeperException, Inte
     while (true) {
 
       MasterLockWatcher masterLockWatcher = new MasterLockWatcher();
-      masterLock = new ZooLock(context.getZooReaderWriter(), zMasterLoc);
+      masterLock =
+          new ZooLock(context.getZooReaderWriter(), zMasterLoc, ZooReader.DISABLED_RETRY_FACTORY);
```
Comment (ctubbsii) [inline server/manager/src/main/java/org/apache/accumulo/master/Master.java:1427]: In these kinds of uses, the code reads as though ZooLock itself won't retry... this is a bit unintuitive, because of the unfortunate name.

File: core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooReader.java (hunk 2 above)
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooReader.java:47]: Looking at the interfaces they are chained together such that they are needed.

**Comment by milleruntime** [comment]
FYI the original ticket #1086 was for the 1.9 branch which will probably be a very different fix.  We may chose to not fix it in 1.10 but just wanted to give a heads up that this could auto close that ticket.

**Comment by dlmarion** [comment]
@milleruntime - good catch. I created another PR for the 1.9 branch at https://github.com/apache/accumulo/pull/1679

File: server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java [hunk 5]
```diff
@@ -1384,4 +1385,14 @@ public final RateLimiter getMajorCompactionWriteLimiter() {
...
+
+  /**
+   * For the tablet server we don't want ZooReaderWriter to keep retrying after a error
+   * communicating with ZooKeeper.
+   */
+  @Override
+  protected ServerContext createServerContext(SiteConfiguration siteConfig) {
+    return new ServerContext(siteConfig, false);
+  }
+
```
Comment (ctubbsii) [inline server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java:1397]: I worry that this will have a larger impact on the various ways ZooKeeper is used in a TabletServer, rather than be a narrow fix for ZooLock. The ServerContext is universal for the process. The only ZooReaderWriter instance for which we want to avoid retries is the one used by ZooLock. Is there a reason we need to change the behavior across the entire tserver?

File: core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooReader.java [hunk 6]
```diff
@@ -38,24 +38,34 @@
...
 
   protected final String keepers;
   protected final int timeout;
+  protected final RetryFactory retryFactory;
 
   public ZooReader(String keepers, int timeout) {
+    this(keepers, timeout, true);
+  }
+
+  public ZooReader(String keepers, int timeout, boolean enableRetries) {
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooReader.java:57]: Could pass in the RetryFactory rather than have a boolean param.

File: core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooReaderWriter.java [hunk 7]
```diff
@@ -46,16 +46,32 @@
...
+    return new ZooReaderWriter(keepers, timeoutInMillis, secret, false);
   }
 
   private final byte[] auth;
 
-  public ZooReaderWriter(String keepers, int timeoutInMillis, String secret) {
-    super(keepers, timeoutInMillis);
+  private ZooReaderWriter(String keepers, int timeoutInMillis, String secret,
+      boolean enableRetries) {
+    super(keepers, timeoutInMillis, enableRetries);
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooReaderWriter.java:74]: Could pass in the retry factory rather than pass a boolean around. That would make the code more clear.

File: server/base/src/main/java/org/apache/accumulo/server/ServerContext.java [hunk 8]
```diff
@@ -68,13 +68,22 @@
   private CryptoService cryptoService = null;
 
   public ServerContext(SiteConfiguration siteConfig) {
-    this(new ServerInfo(siteConfig));
+    this(new ServerInfo(siteConfig), true);
+  }
+
+  public ServerContext(SiteConfiguration siteConfig, boolean enableZookeeperRetries) {
+    this(new ServerInfo(siteConfig), enableZookeeperRetries);
```
Comment (ctubbsii) [inline server/base/src/main/java/org/apache/accumulo/server/ServerContext.java:75]: Could pass in the retry factory rather than pass around booleans.

File: server/base/src/main/java/org/apache/accumulo/server/ServerContext.java [hunk 9]
```diff
@@ -68,13 +68,22 @@
...
   private ServerContext(ServerInfo info) {
+    this(info, true);
+  }
+
+  private ServerContext(ServerInfo info, boolean enableRetries) {
     super(SingletonReservation.noop(), info, info.getSiteConfiguration());
     this.info = info;
-    zooReaderWriter = new ZooReaderWriter(info.getSiteConfiguration());
+    zooReaderWriter = enableRetries ? ZooReaderWriter.retriesEnabled(info.getSiteConfiguration())
+        : ZooReaderWriter.retriesDisabled(info.getSiteConfiguration());
```
Comment (ctubbsii) [inline server/base/src/main/java/org/apache/accumulo/server/ServerContext.java:86]: This worries me. It seems outside the scope of ServerContext to be handling this kind of thing. ServerContext should merely be a holder of the ZRW that is available to the server for updating/reading from ZK. The affected code is in ZooLock, and is not a problem with ZooReaderWriter's retry behavior more generally.

Comment (keith-turner) [reply server/base/src/main/java/org/apache/accumulo/server/ServerContext.java:86]: I am not sure this a good idea, but its something I thought of after reading this comment.  Could make every ZooReaderWriter method accept a retry factory and have overloaded versions that do not take a retry factory and use a default.  This  seems tedious though, but its the best I could think of so far to associate the retry logic with method calls instead of the object.

Comment (ctubbsii) [reply server/base/src/main/java/org/apache/accumulo/server/ServerContext.java:86]: If we do something like add a retry to every method, I'd like to revisit my effort to clean up the ZooReaderWriter stuff, like what I had started to do in #1459 ; it would be easier to add something like a retry to every method, if the code were cleaned up a bit first. I can revisit my previous cleanup effort over the next week.

In the meantime, I still think it's probably best to simply avoid the problematic ZRW code in the ZooLock. ZooLock does not need to use the general-purpose utility code, when it can use something specific to its needs.

Comment (dlmarion) [reply server/base/src/main/java/org/apache/accumulo/server/ServerContext.java:86]: Thanks for the feedback on the PR. I read back through the comments on #1086 and the comments above. My previous attempts were to create a code path through the existing objects that didn't retry instead of creating something new that also talks with ZK. It sounds as if you both might be leaning toward a solution that does not use ZooLock for the tablet server lock, but uses something else that does not retry and communicates directly with ZooKeeper. Is that correct?

Comment (dlmarion) [reply server/base/src/main/java/org/apache/accumulo/server/ServerContext.java:86]: @ctubbsii - did you want me to wait for #1459 to be completed before revisiting this?

Comment (ctubbsii) [reply server/base/src/main/java/org/apache/accumulo/server/ServerContext.java:86]: > It sounds as if you both might be leaning toward a solution that does not use ZooLock for the tablet server lock, but uses something else that does not retry and communicates directly with ZooKeeper. Is that correct?

I don't think that's quite right. I don't think it makes sense to distance ourselves from using ZooLock for locking (doesn't it exist for that purpose?). Rather, I think we're suggesting that the implementation of ZooLock should be changed to distance itself from the general-purpose utilities of ZooReaderWriter/ZooSession. This can be done by making ZooLock never use those at all and use something of its own, or as Keith suggested, adding special-purpose methods on those utilities for ZooLock to safely use.

> @ctubbsii - did you want me to wait for #1459 to be completed before revisiting this?

That depends. #1459 had no intention of touching ZooLock, so if you want to try to make ZooLock work without ZooReaderWriter/ZooSession, then your efforts should not conflict with that and can proceed independently. If, however, you want to try to add special-purpose methods for ZooLock on ZooReaderWriter/ZooSession, then it should wait, and I can prioritize my efforts next week to clean those up to make it easier for you.

Comment (dlmarion) [reply server/base/src/main/java/org/apache/accumulo/server/ServerContext.java:86]: > Rather, I think we're suggesting that the implementation of ZooLock should be changed to distance itself from the general-purpose utilities of ZooReaderWriter/ZooSession. This can be done by making ZooLock never use those at all and use something of its own, or as Keith suggested, adding special-purpose methods on those utilities for ZooLock to safely use.

Understood, I'll work down that path. Thanks.

Comment (keith-turner) [reply server/base/src/main/java/org/apache/accumulo/server/ServerContext.java:86]: @dlmarion  I was thinking of looking into making ZooLock use Zookeeper directly to see if that is workable.

**Comment by ctubbsii** [comment]
I have completed a refactor of some of the Zoo-related utilities in #1755 . After that PR is merged, it might be easier to identify alternate ways to address this. (Or, it may not be any help... I'm not sure. It's been awhile since I looked at this issue.)

**Comment by ctubbsii** [comment]
@dlmarion This was closed as marked as "Done" for 2.1.0, but I don't think it was merged. I think it's being addressed via a different approach. Is that correct?

**Comment by dlmarion** [comment]
I delete the previous PR and branch as it's being addressed [here](https://github.com/dlmarion/accumulo/pull/3). It looks like `automation` marked it `Done`.
